server/scsynth/SC_PortAudio.cpp

```diff
@@ -262,39 +268,198 @@ int SC_PortAudioDriver::PortAudioCallback(const void* input, void* output, unsig
+        }
+        if (err != paNoError) {
+            fprintf(stdout, "PortAudio error: %s\nRequested sample rate %f for device %s is not supported\n",
+                    Pa_GetErrorText(err), sampleRate, Pa_GetDeviceInfo(*device)->name);
+            return err;
+        }
+    }
+    // in case we still don't have a proper device, use the default device
+    if (*device == paNoDevice) {
+        if (isInput) {
+            *device = Pa_GetDefaultInputDevice();
+        } else {
+            *device = Pa_GetDefaultOutputDevice();
+        }
+
+        if (*device != paNoDevice)
+            fprintf(stdout, "Selecting default system %s device\n", (isInput ? "input" : "output"));
+    }
+    return paNoError;
+}
+
+void SC_PortAudioDriver::SelectMatchingPaDevice(int* matchingDevice, int* knownDevice, IOType matchingDeviceType) {
+    if (*matchingDevice == paNoDevice && *knownDevice != paNoDevice) {
+        const PaHostApiInfo* apiInfo;
+        apiInfo = Pa_GetHostApiInfo(Pa_GetDeviceInfo(*knownDevice)->hostApi);
+
+        bool isInput;
+        if (matchingDeviceType == IOType::Input)
+            isInput = true;
+        else if (matchingDeviceType == IOType::Output)
+            isInput = false;
+
+        if (isInput) {
+            *matchingDevice = apiInfo->defaultInputDevice;
+        } else {
+            *matchingDevice = apiInfo->defaultOutputDevice;
         }
+        if (*matchingDevice != paNoDevice)
+            fprintf(stdout, "Selecting default %s %s device\n", apiInfo->name, (isInput ? "input" : "output"));
     }
 }
 
+// this function will select default PA devices if they are not defined
+// it will also try to check for some configuration problems
+// numIns, numOuts and sampleRate are only the requested values, can change later
+PaError SC_PortAudioDriver::CheckPaDevices(int* inDevice, int* outDevice, int numIns, int numOuts, double sampleRate) {
+    // make independent checks whether we are using only input, only output, or both input and outputTouched
+    if (numIns && !numOuts) {
+        // inputs only
+        *outDevice = paNoDevice;
+        // check for requested sample rate or select the default device
+        return CheckSinglePaDevice(inDevice, sampleRate, IOType::Input);
+    } else if (!numIns && numOuts) {
+        // outputs only
+        *inDevice = paNoDevice;
+        // check for requested sample rate or select the default device
+        return CheckSinglePaDevice(outDevice, sampleRate, IOType::Output);
+    } else if (numIns && numOuts) {
+        // inputs and outputs
+        // if one device is specified, let's try to open another one on matching api
+        // try matching input to output
+        SelectMatchingPaDevice(inDevice, outDevice, IOType::Input);
+        // then try matching output to input
+        SelectMatchingPaDevice(outDevice, inDevice, IOType::Output);
+        // check if devices are having mismatched API, but only if they are defined
```

Comment:
I removed some of them. 

My intent was to indicate logic in choosing devices, since there is more steps to it then just "give me selected device or the default one". I can see what you say, but I don't know what's the right balance here. I'm happy to remove other comments if you think that's optimal.

Reply:
the only comment here that is useful IMO is `// if one device is specified, let's try to open another one on matching api`, the rest just gets in the way of understanding and can be easily figured out from context. Good comments should "tell why, not how".